### ISSUE TITLE
Health-Based Bar

### ISSUE THREAD

**Issue by dctrlight** [danielkrupinski/Osiris#1226]
Having an option to make the health bar color health based would be cool

**Comment by dapaster** [comment]
Agree

**Comment by AbdallahMehiz** [comment]
> Having an option to make the health bar color health based would be cool

what does that even mean?

**Comment by dapaster** [comment]
look at the ESP . The heatlh bar color only have manual color and rainbow only

**Comment by jimbob88** [comment]
> 
> 
> > Having an option to make the health bar color health based would be cool
> 
> what does that even mean?

What he means is something along the lines of when they are between 90-100% health make the box green, when it is between 40-89 amber and 0-39 red or something kind of like that.

**Comment by AbdallahMehiz** [comment]
> > > Having an option to make the health bar color health based would be cool
> > 
> > 
> > what does that even mean?
> 
> What he means is something along the lines of when they are between 90-100% health make the box green, when it is between 40-89 amber and 0-39 red or something kind of like that.

ok i get it._.

**Comment by jimbob88** [comment]
```cpp
            const auto currHealth{ ((entity->health()) / static_cast<float>(maxHealth)) * 100 };

            float green_color[3]{ 0.0f, 0.5f, 0.0f };
            float amber_color[3]{ 1.0f, 0.74609375f, 0.0f };
            float red_color[3]{ 1.0f, 0.0f, 0.0f };

            if (config.healthBar.rainbow) {
                interfaces.surface->setDrawColor(rainbowColor(memory.globalVars->realtime, config.healthBar.rainbowSpeed));
            }
            else {
                if (90 <= currHealth && currHealth <= 100) {
                    interfaces.surface->setDrawColor(green_color);
                }
                else if (40 <= currHealth && currHealth <= 89) {
                    interfaces.surface->setDrawColor(amber_color);
                }
                else if (0 <= currHealth && currHealth <= 39) {
                    interfaces.surface->setDrawColor(red_color);
                }
                else {
                    interfaces.surface->setDrawColor(red_color);
                }
            }
```
This is my basic implementation of this concept, the dll is [here](https://drive.google.com/file/d/13hM28uV-hCbkQ7Ibe-pRmpM5tLncHgU7/view?usp=sharing), I used SazInjector to inject the dll.

Hope this helps,
Jim

**Comment by dapaster** [comment]
> 
> 
> ```c++
>             const auto currHealth{ ((entity->health()) / static_cast<float>(maxHealth)) * 100 };
> 
>             float green_color[3]{ 0.0f, 0.5f, 0.0f };
>             float amber_color[3]{ 1.0f, 0.74609375f, 0.0f };
>             float red_color[3]{ 1.0f, 0.0f, 0.0f };
> 
>             if (config.healthBar.rainbow) {
>                 interfaces.surface->setDrawColor(rainbowColor(memory.globalVars->realtime, config.healthBar.rainbowSpeed));
>             }
>             else {
>                 if (90 <= currHealth && currHealth <= 100) {
>                     interfaces.surface->setDrawColor(green_color);
>                 }
>                 else if (40 <= currHealth && currHealth <= 89) {
>                     interfaces.surface->setDrawColor(amber_color);
>                 }
>                 else if (0 <= currHealth && currHealth <= 39) {
>                     interfaces.surface->setDrawColor(red_color);
>                 }
>                 else {
>                     interfaces.surface->setDrawColor(red_color);
>                 }
>             }
> ```
> 
> This is my basic implementation of this concept, the dll is [here](https://drive.google.com/file/d/13hM28uV-hCbkQ7Ibe-pRmpM5tLncHgU7/view?usp=sharing), I used SazInjector to inject the dll.
> 
> Hope this helps,
> Jim

Thanks

**Comment by jimbob88** [comment]
# New release
This has now been released on my github, with the current release being [here](https://github.com/jimbob88/Osiris/releases/tag/v1.3)
# Changes
It has added a check button to enable the feature written [here](https://github.com/danielkrupinski/Osiris/issues/1226#issuecomment-596122527), which is found when pressing on the colour option for a menubar!
Also it includes the [human bhop feature](https://github.com/danielkrupinski/Osiris/issues/1123#issuecomment-588666316) from [this pull request](https://github.com/danielkrupinski/Osiris/pull/1162)

_Feel free to drop me a feature request ;))_

Hope this helps y'all :)
Jim

**Comment by dapaster** [comment]
> 
> # New release
> 
> This has now been released on my github, with the current release being [here](https://github.com/jimbob88/Osiris/releases/tag/v1.3)
> # Changes
> 
> It has added a check button to enable the feature written [here](https://github.com/danielkrupinski/Osiris/issues/1226#issuecomment-596122527), which is found when pressing on the colour option for a menubar!
> Also it includes the [human bhop feature](https://github.com/danielkrupinski/Osiris/issues/1123#issuecomment-588666316) from [this pull request](https://github.com/danielkrupinski/Osiris/pull/1162)
> 
> _Feel free to drop me a feature request ;))_
> 
> Hope this helps y'all :)
> Jim

can you add hands & sleeves into Glow list too

**Comment by Kron1Q** [comment]
> ```c++
>             const auto currHealth{ ((entity->health()) / static_cast<float>(maxHealth)) * 100 };
> 
>             float green_color[3]{ 0.0f, 0.5f, 0.0f };
>             float amber_color[3]{ 1.0f, 0.74609375f, 0.0f };
>             float red_color[3]{ 1.0f, 0.0f, 0.0f };
> 
>             if (config.healthBar.rainbow) {
>                 interfaces.surface->setDrawColor(rainbowColor(memory.globalVars->realtime, config.healthBar.rainbowSpeed));
>             }
>             else {
>                 if (90 <= currHealth && currHealth <= 100) {
>                     interfaces.surface->setDrawColor(green_color);
>                 }
>                 else if (40 <= currHealth && currHealth <= 89) {
>                     interfaces.surface->setDrawColor(amber_color);
>                 }
>                 else if (0 <= currHealth && currHealth <= 39) {
>                     interfaces.surface->setDrawColor(red_color);
>                 }
>                 else {
>                     interfaces.surface->setDrawColor(red_color);
>                 }
>             }
> ```
> 
> This is my basic implementation of this concept, the dll is [here](https://drive.google.com/file/d/13hM28uV-hCbkQ7Ibe-pRmpM5tLncHgU7/view?usp=sharing), I used SazInjector to inject the dll.
> 
> Hope this helps,
> Jim

Thats ugly af. There is Health based function in chams already, just c+p into healthbar.

        if (chams.healthBased && health)
            material->colorModulate(1.0f - health / 100.0f, health / 100.0f, 0.0f);

best regards.

**Comment by jimbob88** [comment]
> 
> 
> > ```c++
> >             const auto currHealth{ ((entity->health()) / static_cast<float>(maxHealth)) * 100 };
> > 
> >             float green_color[3]{ 0.0f, 0.5f, 0.0f };
> >             float amber_color[3]{ 1.0f, 0.74609375f, 0.0f };
> >             float red_color[3]{ 1.0f, 0.0f, 0.0f };
> > 
> >             if (config.healthBar.rainbow) {
> >                 interfaces.surface->setDrawColor(rainbowColor(memory.globalVars->realtime, config.healthBar.rainbowSpeed));
> >             }
> >             else {
> >                 if (90 <= currHealth && currHealth <= 100) {
> >                     interfaces.surface->setDrawColor(green_color);
> >                 }
> >                 else if (40 <= currHealth && currHealth <= 89) {
> >                     interfaces.surface->setDrawColor(amber_color);
> >                 }
> >                 else if (0 <= currHealth && currHealth <= 39) {
> >                     interfaces.surface->setDrawColor(red_color);
> >                 }
> >                 else {
> >                     interfaces.surface->setDrawColor(red_color);
> >                 }
> >             }
> > ```
> > 
> > 
> > This is my basic implementation of this concept, the dll is [here](https://drive.google.com/file/d/13hM28uV-hCbkQ7Ibe-pRmpM5tLncHgU7/view?usp=sharing), I used SazInjector to inject the dll.
> > Hope this helps,
> > Jim
> 
> Thats ugly af. There is Health based function in chams already, just c+p into healthbar.
> 
> ```
>     if (chams.healthBased && health)
>         material->colorModulate(1.0f - health / 100.0f, health / 100.0f, 0.0f);
> ```
> 
> best regards.

Before calling my code "ugly af", make sure to verify the fact your code is actually functional ;)

The issues with your code:
 * You have tried to use the variable, `health`, that doesn't appear in the scope of `renderPlayerBox`, one should instead have used `entity->health()`, this as a  first means you cannot simply `C+P`
 * You have tried to use the function `material`, this doesn't appear in this scope either.
 * Even if `material->colorModulate()` did appear in the scope of `renderPlayerBox`, your code still would not work because the drawing function for `healthBar` and `armorBar` uses `interfaces.surface->drawFilledRect()` instead.

What was useful from your code:
 * The formula  `1.0f - health / 100.0f, health / 100.0f, 0.0f` I have rewritten and applied to my current build of the code: `{ 1.0f - currHealth, currHealth, 0.0f }` so thank you for bringing my attention to that formula :)

I have now updated the code as of [commit c3113e2](https://github.com/jimbob88/Osiris/commit/c3113e2d9a2e68a938162da00aaf55579072e2dc) to include  the afforementioned formula instead of the prior system.

Kind regards,
Jim

**Comment by Kron1Q** [comment]
Dude, WTF. You are trolling me or sth now. I just wanted to show you that there is something like Health based color in this code, and chams as its example, never said its straight c+p, u have to change few things to make it works. And yes, your first code posted above is ugly as fuck. Current one is nice <3.

**Comment by jimbob88** [comment]
> There is Health based function in chams already, just c+p into healthbar.

I do apologise I just thought `just c+p` meant to just copy and paste, sorry there, didn't mean to offend.

Regards,
Jim

**Comment by Kron1Q** [comment]
No problem, I could be more accurate. In addition you can replace this ugly color between green and red with nice full yellow doing something like that:

std::min((2.0f * (100 - Health) / 100.0f), 1.0f) , std::min((2.0f * Health) / 100.0f, 1.0f) , 0.f

But thats my personal preference.

**Comment by jimbob88** [comment]
Yeah you do make a good point the orange is quite dark with the charms setup, thanks for the pointer, 
I'll implement this now :)

Thanks very much,
Jim

**Comment by jimbob88** [comment]
The Health-based bar has now been updated, the source code is [here](https://github.com/jimbob88/Osiris/tree/health_based_bars), and the builds are available [here](https://github.com/jimbob88/Osiris/releases). Kron1Q's fixes are added in `v1.3.1`.

Hope this helps everyone,
Jim

P.S. Thanks to Kron1Q for showing the implementation for fixed colours! 👍
